packages/publish/src/publish.ts

```diff
@@ -105,6 +110,154 @@ function handler(app: Application) {
+    if (!remoteConfiguration) {
+      app.log.info(`publish not configured for (${repoName})`);
+      return;
+    }
+
+    // if missing the label, skip
+    if (
+      !context.payload.pull_request.labels.some(
+        label => label.name === PRE_RELEASE_LABEL
+      )
+    ) {
+      app.log.info(
+        `ignoring non-candidate label action (${context.payload.pull_request.labels.join(
+          ', '
+        )})`
+      );
+      return;
+    }
+
+    // We will publish a current snapshot of the repository:
+    const name = context.payload.repository.name;
+    const owner = context.payload.repository.owner.login;
+    const tarballURL = `https://codeload.github.com/${owner}/${name}/legacy.tar.gz/master`;
+
+    // Create a temporary directory to unpack release tarball to:
+    const unpackPath = handler.unpackPath();
+    app.log.info(`creating tmp directory ${unpackPath}`);
+    await fs.mkdir(unpackPath, {
+      recursive: true,
+    });
+
+    // Unpack the tarball to the directory we just created:
+    await fetch(tarballURL).then(res => {
+      return new Promise((resolve, reject) => {
+        const dest = tar.x({
+          C: unpackPath,
+        });
+        dest.on('error', reject);
+        dest.on('close', () => {
+          app.log.info('finished unpacking tarball');
+          return resolve();
+        });
+        res.body.pipe(dest);
+      });
+    });
+
+    // The tarball most likely had an inner folder, which we traverse
+    // into, before performing the publication step:
+    const files = await fs.readdir(unpackPath, {
+      withFileTypes: true,
+    });
+    let pkgPath = unpackPath;
+    if (files.length === 1 && files[0].isDirectory()) {
+      pkgPath = `${pkgPath}/${files[0].name}`;
+    }
+
+    // Update the package.json on disk with a candidate version number, i.e.,
+    // v3.0.0-beta.0:
+    const candidateVersion = await setCandidateVersion(
+      pkgPath,
+      context.payload.pull_request.title
+    );
+
+    const secret: Secret = await handler.getPublicationSecrets(
+      app,
+      remoteConfiguration
+    );
+    if (secret && secret.payload && secret.payload.data) {
+      // Create a pre-release on GitHub for this release candidate:
+      try {
+        await context.github.repos.createRelease({
+          owner: context.payload.repository.owner.login,
+          repo: context.payload.repository.name,
+          tag_name: `v${candidateVersion.version}`,
+          name: `v${candidateVersion.version} Pre-release`,
+          prerelease: true,
+          target_commitish: 'master',
+          body: `This is a pre-release for testing purposes of #${context.payload.pull_request.number}.\n\nInstall by running \`npm install ${candidateVersion.name}@next\``,
+        });
+      } catch (err) {
+        // The most likely error is that the release already existed, and we
+        // had a publication failure; we can skip an error creating the release:
+        app.log.error(err);
```

Comment:
How do these failures get monitored today?  Can we verify it's the error case we think it is, and otherwise re-throw?  Worried about swallowing errors.

Reply:
@JustinBeckwith 👍 I'll figure out what error code t is and catch specifically.